### PR TITLE
[flat-web] build docker and deployfile

### PR DESCRIPTION
**Pull request by ooeyuna**

### CODE REVIEW

File: config/.env [hunk 1]
```diff
@@ -1,23 +0,0 @@
-NETLESS_APP_IDENTIFIER=
```
Comment (BlackHole1) [inline config/.env]: It should not be deleted

File: deploy/Dockerfile.nginx [hunk 2]
```diff
@@ -0,0 +1,9 @@
+FROM nginx
```
Comment (BlackHole1) [inline deploy/Dockerfile.nginx]: deploy/flat-web/nginx.Dockerfile
or
web/flat-web/deploy/nginx.Dockerfile

File: deploy/deploy-dev.yaml [hunk 3]
```diff
@@ -0,0 +1,85 @@
+apiVersion: apps/v1
```
Comment (BlackHole1) [inline deploy/deploy-dev.yaml]: use helm

File: deploy/deploy-dev.yaml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      maxSurge: 25%
+      maxUnavailable: 25%
+    type: RollingUpdate
+  template:
+    metadata:
+      labels:
+        app: flat-web
+    spec:
+      containers:
+        - image: 'agoraflat/flat-web:dev-06101722'
```
Comment (BlackHole1) [inline deploy/deploy-dev.yaml:26]: agoraflat/flat-web:dev-latest

File: deploy/deploy-dev.yaml (hunk 3 above)
Comment (BlackHole1) [reply deploy/deploy-dev.yaml]: https://github.com/netless-io/flat-server/helm

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: .github/encrypt_secret.sh [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+#   ubuntu/debian: apt-get install -y gpg
+#   centos/RedHat: apt-get install -y gpg
+#
+# Usage:
+#
+# ./.github/envcrypt_secret.sh [development|production]
+
+set -e;
+
+ENV=$1
```
Comment (BlackHole1) [inline .github/encrypt_secret.sh:18]: Check if `$1` exists

File: .github/encrypt_secret.sh [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+if [ -f "config/.env.${ENV}.local" ]
+then
+    echo "encrypt config/.env.${ENV}.local to base64 format"
+    gpg --symmetric  --output tmp --cipher-algo AES256 config/.env.${ENV}.local
```
Comment (BlackHole1) [inline .github/encrypt_secret.sh]:
```suggestion
    gpg --symmetric --output tmp --cipher-algo AES256 config/.env.${ENV}.local
```

File: .github/encrypt_secret.sh [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    cat tmp | base64 -w 0 > config/.env.${ENV}.local.gpg
+    rm tmp
+    echo "encrypt complete, add .env.${ENV}.local.gpg to github secret >>GPG_SECRET_FILE<< and PASSWORD to github >>GPG_SECRET_PASSWORD<< for flat-web docker build"
```
Comment (BlackHole1) [inline .github/encrypt_secret.sh]:
```suggestion
    echo "encrypt complete, add .env.${ENV}.local.gpg to github secret >>GPG_SECRET_FILE<< and PASSWORD to github >>GPG_SECRET_PASSWORD<< for flat docker build"
```

We shouldn't limit this script to `flat-web`, I'd prefer it to be a generic script

File: .github/workflows/docker-build.yml [hunk 8]
```diff
@@ -0,0 +1,92 @@
+name: flat-web Docker Build
+on:
+  push:
+    branches:
+      - "main"
+      - "sa/deploy"
```
Comment (BlackHole1) [inline .github/workflows/docker-build.yml]: remove `sa/deploy`

File: .github/workflows/docker-build.yml [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    paths-ignore:
+      - "docs/**"
+      - "*.md"
+      - ".yarnrc"
+      - "LICENSE"
+      - "desktop"
```
Comment (BlackHole1) [inline .github/workflows/docker-build.yml]:
```yml
paths:
  - "web/flat-web/**"
```

File: config/.env.development [hunk 10]
```diff
@@ -1,27 +0,0 @@
-NETLESS_APP_IDENTIFIER=cFjxAJjiEeuUQ0211QCRBw/mO9uJB_DiCIqug
```
Comment (BlackHole1) [inline config/.env.development]: why delete?

File: .github/workflows/docker-build.yml [hunk 11]
```diff
@@ -0,0 +1,92 @@
+name: flat-web Docker Build
```
Comment (BlackHole1) [inline .github/workflows/docker-build.yml]: filename need modify:

`docker-build.yml` -> `flat-web-deploy.yml`

File: .github/workflows/docker-build.yml [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+      - name: docker cache
+        run: |
+          docker build . -f web/flat-web/deploy/Dockerfile.cache -t agoraflat/flat-web:cache --cache-from agoraflat/flat-web:cache --ulimit nofile=65535:65535
+          docker push agoraflat/flat-web:cache
+
+      - name: docker build and push with commit sha
+        env:
+          ENV: development
+          TAG: development-${{ github.sha }}
+        run: ./web/flat-web/deploy/build.sh && docker push agoraflat/flat-web:$TAG
```
Comment (BlackHole1) [inline .github/workflows/docker-build.yml]:
```suggestion
        run: |
          docker build . -f web/flat-web/deploy/Dockerfile -t agoraflat/flat-web:$TAG --build-arg build_env=development --ulimit nofile=65535:65535
          docker push agoraflat/flat-web:$TAG
```

File: .github/workflows/docker-build.yml [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+      - name: docker build and push latest
+        env:
+          ENV: development
+          TAG: dev-latest
+        run: ./web/flat-web/deploy/build.sh && docker push agoraflat/flat-web:$TAG
```
Comment (BlackHole1) [inline .github/workflows/docker-build.yml]:
```suggestion
        run: |
          docker build . -f web/flat-web/deploy/Dockerfile -t agoraflat/flat-web:$TAG --build-arg build_env=development --ulimit nofile=65535:65535
          docker push agoraflat/flat-web:$TAG
```

File: .github/workflows/docker-build.yml [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+      - name: docker cache
+        run: |
+          docker build . -f web/flat-web/deploy/Dockerfile.cache -t agoraflat/flat-web:cache --cache-from agoraflat/flat-web:cache --ulimit nofile=65535:65535
+          docker push agoraflat/flat-web:cache
+
+      - name: docker build and push with commit sha
+        env:
+          ENV: production
+          TAG: production-${{ github.sha }}
+        run: ./web/flat-web/deploy/build.sh && docker push agoraflat/flat-web:$TAG
```
Comment (BlackHole1) [inline .github/workflows/docker-build.yml]: the same as above

File: .github/workflows/docker-build.yml [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+      - name: docker build and push latest
+        env:
+          ENV: production
+          TAG: latest
+        run: ./web/flat-web/deploy/build.sh && docker push agoraflat/flat-web:$TAG
```
Comment (BlackHole1) [inline .github/workflows/docker-build.yml]: the same as above

File: .github/workflows/docker-build.yml [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+          TAG: development-${{ github.sha }}
+        run: ./web/flat-web/deploy/build.sh && docker push agoraflat/flat-web:$TAG
+
+      - name: docker build and push latest
+        env:
+          ENV: development
+          TAG: dev-latest
+        run: ./web/flat-web/deploy/build.sh && docker push agoraflat/flat-web:$TAG
+
+  build_prod:
```
Comment (BlackHole1) [inline .github/workflows/docker-build.yml]: Why do we deploy dev and prod at the same time when we deploy?

File: web/flat-web/deploy/build.sh [hunk 17]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
```
Comment (BlackHole1) [inline web/flat-web/deploy/build.sh]: delete file

File: web/flat-web/deploy/nginx.conf [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+        default_type text/css;
+    }
+
+    location ~* ^.+\.js$ {
+        default_type text/javascript;
+    }
+
+    # location ~* ^.+\.svg$ {
+    #     default_type image/svg+xml;
+    # }
```
Comment (BlackHole1) [inline web/flat-web/deploy/nginx.conf]: remove

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: .github/workflows/docker-build.yml (hunk 16 above)
Comment (ooeyuna) [reply .github/workflows/docker-build.yml]: maybe we need build production image at another workflow

**Pull request closed without merging**